Carousel: Refactor to adhere to new design

barklund opened the pull request:
## Summary

This is a complete overhaul of the carousel components (code-wise) as well as adding a new structure for it.

I decided to update the thumbnail highlight to the new design (with a full thin border rather than just a fat top border) as the old design looked weird with the new changes.

## Relevant Technical Choices

<!-- Please describe your changes. -->

## To-do

* [x] Refactor carousel
* [x] Change thumbnail sizing and layout
* [x] Retire old compact indicator
* [x] Change scroll arrow behavior
* [x] Change carousel menu layout
* [x] Update thumbnail highlight to new design
* [x] Make sure tests still run

## User-facing changes

<!-- Please describe your changes. -->

## Testing Instructions

* Make sure thumbnail carousel, grid view and keyboard shortcuts still work as intended.

---

<!-- Please reference the issue(s) this PR addresses. -->

Fixes #5692
Fixes #5077

Commit:
Moved carousel to new component folder along with relevant files

Commit:
Refactored carousel into many components

Commit:
Fix carousel menu icons

Commit:
Fixed grid view page preview reference

Commit:
Almost fixed page borders - old style clashed with round corners

Commit:
Fixed semantics of elements

Commit:
Fix all the ref listeners - must re-run when ref is actually populated

Commit:
Fixed rearrange bug

Commit:
Merge branch 'main' into add/5692-new-carousel

# Conflicts:
#	assets/src/dashboard/utils/index.js
#	assets/src/design-system/components/keyboard/gridview/test/getIndex.js
#	assets/src/edit-story/components/canvas/carousel/index.js
#	assets/src/edit-story/components/canvas/gridview/test/useGridViewKeys.js
#	assets/src/edit-story/components/carousel/gridview/test/useGridViewKeys.js
#	assets/src/edit-story/components/carousel/gridview/useGridViewKeys.js
#	assets/src/edit-story/components/carousel/useCarouselKeys.js
#	assets/src/edit-story/components/keyboard/index.js

miina commented on assets/src/edit-story/components/keyboard/index.js:
```diff
@@ -0,0 +1,403 @@
+/*
+ * Copyright 2020 Google LLC
```
Isn't all this a duplicate from `design-system/components/keyboard/index.js`? Shouldn't it be used directly from there instead? Or what's the idea of having it in both places? Thoughts?

swissspidy replied:
Looks like a leftover file from fixing merge conflicts after my refactoring.

barklund replied:
Ah, I thought I fixed all the merge problems - will delete this too.

barklund replied:
Removed in eb983d560.

miina commented on assets/src/edit-story/components/carousel/useCarouselSizing.js:
```diff
@@ -0,0 +1,47 @@
+function useCarouselSizing({ availableSpace, numPages }) {
+  return useMemo(() => {
+    const isWideWorkspace = availableSpace >= 700;
+    const spaceForCarousel = availableSpace - 2 * 167 - 2 * 32;
```
Can we use constants here instead of magic numbers?

barklund replied:
Haha, yes, I was going to do that. They're constants elsewhere, will move them to a central location.

barklund replied:
Fixed in 26ccacdc4.

miina commented on assets/src/edit-story/components/carousel/useCarouselScroll.js:
```diff
@@ -0,0 +1,91 @@
+import { useConfig } from '../../app';
+
+function useCarouselScroll({ listElement, carouselWidth, hasOverflow }) {
+  const [scrollPercentage, setScrollPercentage] = useState(0);
```
Nit: this value doesn't seem to be in percentage actually. Not sure what would be a better name either so unless you have a more accurate one, can probably stay as is.

barklund replied:
`scrollRatio`?

barklund replied:
Fixed in b1d4e1472.

miina commented on assets/src/edit-story/components/carousel/gridview/index.js:
```diff
@@ -321,7 +321,7 @@ function GridView() {
+                    page={page}
                     width={width}
                     height={height}
                     dragIndicatorOffset={GRID_GAP / 2}
```
Is this used anywhere btw?

barklund replied:
No idea, will check.

barklund replied:
Was unused - removed in 57ef174a1.

miina commented on assets/src/edit-story/components/carousel/carouselProvider.js:
```diff
@@ -0,0 +1,139 @@
+      numPages,
+      currentPageId,
+      canScrollBack,
+      canScrollForward,
```
Left a separate comment as well but if both `canScroll*`s are `false`, should the buttons be displayed at all?

barklund replied:
Hm, the design isn't clear on that. I'll ask, thanks!

barklund replied:
Forwarded to Sam

barklund replied:
We'll hide completely if no overflow. Fixed in d05008525.

swissspidy commented on assets/src/edit-story/components/carousel/carouselProvider.js:
```diff
@@ -0,0 +1,139 @@
+    currentPageRef.scrollIntoView({
+      inline: 'center',
+      behavior: 'smooth',
+    });
```
While we're at it, might be a good time to address #5077 by using `scrollBy` with an offset instead of `scrollIntoView`.

Is that easily possible?

barklund replied:
Hm, it should be. I should be able to calculate all the numbers quite easily.

barklund replied:
Done in 72843dd80 - and it looks freaking nice, if I can say so myself. I actually [ended up using `scrollTo` rather than `scrollBy`](https://github.com/google/web-stories-wp/pull/5941/files#diff-60efa86d6d8735e06b04393cb877f4fa0242b75724db8c071b4d135da09b2692R100-R153) though 🤷

Commit:
Removed unused file

Commit:
Moved constants to separate file

Commit:
Renamed prop

Commit:
Removed unused prop

Commit:
Simplified function

Commit:
Implemented much cleaner scroll fix - and looks bloody nice!

Commit:
Re-add label for tests

Commit:
Fixed esc2blur for direct node references

Commit:
Fixed re-setting focus to new active page, when a page is deleted

Commit:
Reverted stuff that removed arrow-navigation focus

Commit:
Hidden arrows when no scroll is available

Commit:
Fixed some karma test

I have no idea why this started failing, but it did so consistently. The handle was not clickable in the top left corner at (1,1), so should never have worked?
